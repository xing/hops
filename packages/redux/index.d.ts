import { Middleware, ReducersMapObject, Action as ActionCreator } from 'redux';
import { ThunkAction } from 'redux-thunk';

declare module 'hops-redux' {
  interface HopsReduxActionCreator {
    action: ActionCreator | ThunkAction<any, any, any, any>;
    path: string;
    exact?: boolean;
    strict?: boolean;
  }

  export interface HopsReduxOptions {
    redux: {
      reducers: ReducersMapObject;
      middlewares?: Middleware[];
      actionCreators?: HopsReduxActionCreator[];
      alwaysDispatchActionsOnClient?: boolean;
    };
  }
}
