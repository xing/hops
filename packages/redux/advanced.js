var React = require('react');
var ReactRouter = require('react-router-dom');
var withSideEffect = require('react-side-effect');
var createLocation = require('history/LocationUtils').createLocation;

var HopsRedux = require('./basic');

var Context = HopsRedux.Context;

var Provider = ReactRouter.withRouter(
  withSideEffect(
    function reducePropsToState (propsList) {
      return propsList[propsList.length - 1];
    },
    function handleStateChangeOnClient (props) {
      if (props) {
        props.setLocation(props.location);
        props.dispatchAll();
      }
    },
    function mapStateOnServer (props) {
      return undefined;
    }
  )(function (props) {
    return props.children ? React.Children.only(props.children) : null;
  })
);

exports.Context = exports.createContext = Context.extend({
  initialize: function (options) {
    Context.prototype.initialize.call(this, options);
    this.actionCreators = options.actionCreators || [];
  },
  bootstrap: function () {
    return this.request ? this.dispatchAll() : Promise.resolve();
  },
  dispatchAll: function () {
    var store = this.getStore();
    var location = this.getLocation();
    var request = this.request;
    return Promise.all(this.actionCreators.map(function (createAction) {
      return store.dispatch(createAction(location, request));
    }));
  },
  setLocation: function (location) {
    this.location = location;
  },
  getLocation: function () {
    return this.location || createLocation(this.request.path);
  },
  enhanceElement: function (reactElement) {
    return Context.prototype.enhanceElement.call(this, React.createElement(
      Provider,
      {
        dispatchAll: this.dispatchAll.bind(this),
        setLocation: this.setLocation.bind(this)
      },
      reactElement
    ));
  }
});

exports.render = function (reactElement, context) {
  if (!(context instanceof exports.Context)) {
    context = new exports.Context(context);
  }
  return HopsRedux.render(reactElement, context);
};
