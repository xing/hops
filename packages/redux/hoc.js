var ReactRouter = require('react-router-dom');
var ReactRedux = require('react-redux');
var withSideEffect = require('react-side-effect');

exports.withLocalAction = function (Component, createAction) {
  return ReactRouter.withRouter(
    ReactRedux.connect(
      function mapStateToProps (state, ownProps) {
        return ownProps;
      },
      function mapDispatchToProps (dispatch, ownProps) {
        return {
          act: function () {
            return dispatch(createAction(ownProps.location));
          }
        };
      }
    )(Component)
  );
};

exports.withAutomaticLocalAction = function (Component) {
  return exports.withLocalAction(
    withSideEffect(
      function reducePropsToState (propsList) {
        return propsList[propsList.length - 1];
      },
      function handleStateChangeOnClient (props) {
        return props && props.act();
      },
      function mapStateOnServer (props) {
        return undefined;
      }
    )(Component)
  );
};
