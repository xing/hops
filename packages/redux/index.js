'use strict';

var React = require('react');
var Redux = require('redux');
var ReactRedux = require('react-redux');
var ReduxThunkMiddleware = require('redux-thunk').default;

var Context = require('hops-react').Context;

var REDUX_STATE = 'REDUX_STATE';

exports.Context = exports.createContext = Context.mixin({
  initialize: function (options) {
    this.reducers = {};
    Object.keys(options.reducers || {}).forEach(
      function (key) {
        this.registerReducer(key, options.reducers[key]);
      }.bind(this)
    );
  },
  registerReducer: function (namespace, reducer) {
    this.reducers[namespace] = reducer;
    if (this.store) {
      this.store.replaceReducer(Redux.combineReducers(this.reducers));
    }
  },
  getStore: function () {
    if (module.hot) {
      this.store = global.store || (global.store = this.createStore());
    }
    return this.store || (this.store = this.createStore());
  },
  createStore: function () {
    return Redux.createStore(
      Redux.combineReducers(this.reducers),
      global[REDUX_STATE],
      this.createEnhancer()
    );
  },
  createEnhancer: function () {
    return this.composeMiddlewares();
  },
  composeEnhancers: function () {
    var compose = global.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || Redux.compose;
    return compose.apply(null, arguments);
  },
  composeMiddlewares: function () {
    return this.composeEnhancers.apply(this, this.applyMiddlewares());
  },
  applyMiddlewares: function () {
    return this.getMiddlewares().map(function (middleware) {
      return Redux.applyMiddleware(middleware);
    });
  },
  getMiddlewares: function () {
    return [ReduxThunkMiddleware];
  },
  enhanceElement: function (reactElement) {
    return React.createElement(
      ReactRedux.Provider,
      {
        store: this.getStore()
      },
      reactElement
    );
  },
  getTemplateData: function (templateData) {
    return Object.assign(templateData, {
      globals: templateData.globals.concat([{
        name: REDUX_STATE,
        value: this.getStore().getState()
      }])
    });
  }
});
