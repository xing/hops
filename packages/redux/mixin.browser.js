const React = require('react');
const {
  compose,
  combineReducers,
  createStore,
  applyMiddleware,
} = require('redux');
const ReduxThunkMiddleware = require('redux-thunk').default;
const { Provider } = require('react-redux');
const {
  Mixin,
  strategies: {
    sync: { override },
  },
} = require('hops-mixin');

const ReduxCompose = global.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

class ReduxMixin extends Mixin {
  constructor(config, element, { redux: options = {} } = {}) {
    super(config);
    this.middlewares = options.middlewares;
    this.reducers = options.reducers || {};
  }

  createStore() {
    return createStore(
      combineReducers(this.reducers),
      global.REDUX_STATE,
      this.composeEnhancers(...this.applyMiddlewares())
    );
  }

  getReduxStore() {
    if (this.store) {
      return this.store;
    }

    if (module.hot) {
      this.store = global.store || (global.store = this.createStore());
      this.store.replaceReducer(combineReducers(this.reducers));
    } else {
      this.store = this.createStore();
    }

    return this.store;
  }

  getReduxMiddlewares() {
    return [
      ReduxThunkMiddleware.withExtraArgument({
        config: this.config,
      }),
    ];
  }

  applyMiddlewares() {
    const middlewares = this.middlewares || this.getReduxMiddlewares();

    return middlewares.map(m => applyMiddleware(m));
  }

  composeEnhancers(...enhancers) {
    return ReduxCompose(...enhancers);
  }

  enhanceElement(reactElement) {
    return React.createElement(
      Provider,
      { store: this.getReduxStore() },
      reactElement
    );
  }
}

ReduxMixin.strategies = {
  getReduxStore: override,
  getReduxMiddlewares: override,
};

module.exports = ReduxMixin;
