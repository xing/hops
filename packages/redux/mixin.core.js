const { Mixin } = require('hops-mixin');

module.exports = class ReduxMixin extends Mixin {
  enhanceTestElement([element, config]) {
    const React = require('react');
    const { Provider } = require('react-redux');
    const { reduxProviderProps = null } = config;
    return [React.createElement(Provider, reduxProviderProps, element), config];
  }
};
