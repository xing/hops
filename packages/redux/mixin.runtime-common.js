const { applyMiddleware } = require('redux');
const ReduxThunkMiddleware = require('redux-thunk').default;
const {
  Mixin,
  strategies: {
    sync: { override },
  },
} = require('hops-mixin');

class ReduxRuntimeCommonMixin extends Mixin {
  constructor(config, element, { redux: options = {} } = {}) {
    super(config);
    this.middlewares = options.middlewares || [];
  }

  getReduxMiddlewares() {
    return [
      ...this.middlewares,
      ReduxThunkMiddleware.withExtraArgument({
        config: this.config,
      }),
    ];
  }

  applyMiddlewares() {
    const middlewares = this.getReduxMiddlewares();

    return middlewares.map(m => applyMiddleware(m));
  }
}

ReduxRuntimeCommonMixin.strategies = {
  getReduxMiddlewares: override,
};

module.exports = ReduxRuntimeCommonMixin;
