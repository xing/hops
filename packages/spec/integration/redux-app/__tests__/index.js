/**
 * @jest-environment jsdom
 */

import React from 'react';
import { createStore, combineReducers } from 'redux';
import testRenderer from 'react-test-renderer';
import { HopsTest } from 'hops-test';

import { App } from '../';

describe('Hops Redux', () => {
  it('should render the App correctly', () => {
    const greeting = () => ({ subject: 'test' });
    const store = createStore(combineReducers({ greeting }));
    const reduxProviderProps = { store };
    const app = testRenderer.create(
      <HopsTest config={{ reduxProviderProps }}>
        <App />
      </HopsTest>
    );

    expect(app).toMatchSnapshot();
  });
});
