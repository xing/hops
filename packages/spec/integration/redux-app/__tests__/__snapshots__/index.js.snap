// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hops Redux should render the App correctly 1`] = `
<h1>
  hello 
  test
  .
</h1>
`;
